enclave/core/sgx/exit.c

```diff
@@ -0,0 +1,169 @@
+    oe_ocall_context_t* context = NULL;
+    td_t* td = oe_get_td();
+
+    if (code == OE_CODE_OCALL)
+    {
+        // Set the last_sp value to any value within this frame.
+        // When ocall returns, it will set RSP to last_sp -  ABI_RED_ZONE and
+        // resume execution.
+        td->base.last_sp = (uint64_t)(void*)&td;
+
+        // Fetch the frame pointer of this function which gives the
+        // rbp, return-address of the caller. Pass it along to
+        // oe_notify_nested_exit_start which will install these values into
+        // the callsite information. The debugger will read these values
+        // and stitch the ocall stack back to this function's caller.
+        asm volatile("mov %%rbp, %0" : "=r"(context));
+        oe_notify_nested_exit_start(arg1, context);
+    }
+    else if (code == OE_CODE_ERET)
+    {
+        if (result != OE_REENTRANT_ECALL)
+        {
+            // Clear out the last_sp.
+            td->base.last_sp = 0;
+        }
+        else
+        {
+            // The rentrant call should not clear out the last_sp value that
+            // any active OCALL may need.
+        }
+
+        // An abort is a short-circuited ecall-exit.
+        // depth must be reset.
+        // Note: It would be good to clear td correctly for abort.
+        if (arg2 == OE_ENCLAVE_ABORTING)
+            td->depth = 0;
+    }
+
+    // Cleanup extended processor states.
+    oe_cleanup_xstates();
+
+    // ENCLU_(EEXIT) must set RDI and RSI to arg1 and arg2.
+    register uint64_t rdi __asm__("rdi") = arg1;
+    register uint64_t rsi __asm__("rsi") = arg2;
+
+    // RBX must contain the host return address.
+    register uint64_t rbx __asm__("rbx") = host_rip;
+
+    // RAX must contin the leaf ENCLU_EEXIT.
+    // For simulation mode, RAX must be 1.
+    register uint64_t rax __asm__("rax") = td->simulate ? 1 : ENCLU_EEXIT;
+
+    // RSP and RBP must be restored to host values. The values are passed in
+    // to the assembly snippet in RCX and RDX.
+    register uint64_t rcx __asm__("rcx") = host_rsp;
+    register uint64_t rdx __asm__("rdx") = host_rbp;
+
+    asm volatile(
+        // Clear general purpose registers.
+        "xor %%r8, %%r8 \n\t"
+        "xor %%r9, %%r9 \n\t"
+        "xor %%r10, %%r10 \n\t"
+        "xor %%r11, %%r11 \n\t"
+        "xor %%r12, %%r12 \n\t"
+        "xor %%r13, %%r13 \n\t"
+        "xor %%r14, %%r14 \n\t"
+        "xor %%r15, %%r15 \n\t"
+
+        // Clear flags.
+        "push %%r15 \n\t"
+        "popf \n\t"
+
+        // Restore host RSP and RBP.
+        "mov %%rcx, %%rsp \n\t"
+        "mov %%rdx, %%rbp \n\t"
+        "xor %%rcx, %%rcx \n\t"
+        "xor %%rdx, %%rdx \n\t"
+
+        // Execute either ENCLU or jmp based on RAX value
+        // which will be 1 for simulation mode. See above.
+        "cmp $1, %%rax \n\t"
+        "je 2f \n\t"
+        "1: enclu \n\t"
+        "2: jmp *%%rbx \n\t"
+        "ud2 \n\t"
```

Comment:
I don't think adding `ud2` here really adds much benefit. Maybe an assert at the end of the function for debug builds only would be sufficient.